7844 codemeta schema

poikilotherm opened the pull request:
**What this PR does / why we need it**:
This is adding the CodeMeta Schema as a default out of the box schema for (new) installations.
This pull request is a first step. Please see the discussion points below for your review. We need to be careful about the scope of this first step to keep compatibility in mind. (There is no schema migration present in the Dataverse application, so when changing data types etc, we need to write SQL database migrations manually!)

**TODOs**

- [X] Test TSV, make screenshots
- [x] Add release note
- [ ] Sort out remaining questions (see below)

**Which issue(s) this PR closes**:

Closes #7844 

**Special notes for your reviewer**:
1. With the new feature of ["metadata block facets" per collection](https://guides.dataverse.org/en/latest/api/native-api.html#set-metadata-block-facets-for-a-dataverse-collection) should we use a different `displayName` for the block? (It currently is "Software Metadata (CodeMeta 2.0)")
1. Should we use the [W3C proposed vocabulary for `applicationCategory`](https://www.w3.org/wiki/WebSchemas/SoftwareApplicationSchema)?
     - Should we go ahead and add `ResearchApplication` to this list and reach out to schema.org and CodeMeta people to push for adding it to the list? (Maybe Google, too?)
     - Should we go ahead and reach out to CodeMeta about a field on scientific method used in the software? (Not covered by `subject` field, which is very coarse anyway)
2. Should we make the `*Requirements` fields use integer values of byte? kilobyte? megabyte? (or similar for CPU) instead of arbitrary text values?
3. Do we want to add docs about the crosswalk of "Dataverse Metadata" to "CodeMeta" to the guides?
4. What other docs do we want to include?
5. Do we want to add https://github.com/SoftwareUnderstanding/software_types (which would extend this beyond pure CodeMeta)
6. Do we want to add a field to allow documenting computational methods in use?
   - There is no standard, vocabulary, schema or ontology for this yet, we'd be on our own. 
   - This might as well be done via Citation Blocks Keywords
   - We could leave this for a later extension of the block

**Suggestions on how to test this**:

- Load the TSV via the usual API call.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

On dataset creation:
![image](https://user-images.githubusercontent.com/5468333/201084087-86623660-34d7-4b32-8ff0-9ef612b49270.png)

On dataset editing:
![image](https://user-images.githubusercontent.com/5468333/201084461-0d64e623-e88a-4d65-a694-d3a5c934f2f5.png)

As JSON-LD export:
![image](https://user-images.githubusercontent.com/5468333/201086726-c03f03cb-bee3-4402-9e4f-08674c91d966.png)

**Is there a release notes update needed for this change?**:

- Yet to be done, as review/extension/discussion needed.

**Additional documentation**:

- Block as TSV: https://docs.google.com/spreadsheets/d/1MsJifbLeRYCdFUXPAOh-KIphgTaxPsc6ICxtmEj_sb4/edit#gid=1781064623
- Mapping CodeMeta to Dataverse software metadata fields: https://docs.google.com/spreadsheets/d/1zcOm1PX2_HTMacgc-sn8aSKTIQBtoaSAAKEPXmKVt3A/edit#gid=0

**Tagging** @doigl @atrisovic @4tikhonov @jggautier @djbrooke @pdurbin (I don't know the GH names of the other WG members)

jggautier commented:
This is great! I'd like to point out two issues that I think are most pressing and I hope could be resolved before this is merged:

1. I think there's a stray tab in line 11, applicationCategory, that's splitting its displayName into two columns:

   <img width="706" alt="Screen Shot 2021-05-17 at 10 55 40 AM" src="https://user-images.githubusercontent.com/18374574/118509962-84bf2580-b6fe-11eb-9a3e-67110c7c1c6f.png">

2. There are a few different ways that the Citation metadatablock's fields are still designed to describe data as opposed to software. It looks like we'll be tackling these issues in future work (such as how metadata is exported), but I hope some of the issues that users will see when depositing software can be resolved before this PR is merged:
   - The tooltips for most of the fields, even fields that make sense for describing software, such as Title, include the word "Dataset".

     One solution might be to abstract the text in the tooltips of the fields in the Citation metadatablock, for example by replacing the word "Dataset" with "deposit".
   - Some of the fields wouldn't make sense for describing software at all, such as "Series", "Date of collection" and "Type of data". If someone is depositing software, I would think they wouldn't need to see these fields.

     To prevent depositors of software from seeing fields that they wouldn't need, one solution might be to recommend that repositories use a Dataverse collection only for software deposits, and when setting up that collection they should hide the fields in the Citation metadatablock that describe data (e.g. "Data of collection", and "Type of data") and enable the Software metadatablock for that Dataverse collection. So repositories, especially "self curated" ones, should not have users deposit a mix of datasets and software into the same Dataverse collection because there wouldn't be a way for the software to know if what the user is depositing is data or software, so the software has no way of showing the relevant metadata fields.

poikilotherm commented:
I once had the idea to actually make the citation block pluggable, non-mandatory. I know this requires A LOT, but maybe it's a way to go, if we don't want other archictural changes like abstracting the concept of sets.

poikilotherm commented:
@doigl has some data about the candidates for `displayOnCreate` from [DaRuS](https://darus.uni-stuttgart.de):

[![grafik](https://user-images.githubusercontent.com/5468333/118640140-ed7ad080-b7d8-11eb-87e1-95a78aa74686.png)](https://dataversecommunity.slack.com/archives/C01F1NLPHU0/p1621330774000900?thread_ts=1621265307.005300&cid=C01F1NLPHU0)

I agree on all of those, except for `applicationCategory`, which has been used not with the vocabulary from W3C but free text. I still think we should not do that.

poikilotherm commented:
There is a [list of programming languages in WikiData](https://w.wiki/uG3), containing ~1500 entries. (Via https://en.wikiversity.org/wiki/Research_in_programming_Wikidata/Programming_languages)

There is an [extensive list of operating systems (not names alone) in WikiData](https://query.wikidata.org/#%23added%202017-03%0A%23List%20of%20%60instances%20of%60%20%22operating%20system%22%20%0ASELECT%20%3Fos%20%3FosLabel%0AWHERE%0A%7B%0A%20%20%20%20%3Fos%20wdt%3AP31%20wd%3AQ9135.%0A%20%20%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22en%22%20%7D%0A%7D) with ~1100 entries. (Via https://en.wikiversity.org/wiki/Research_in_programming_Wikidata/Operating_systems)

We might wanna play with the OS query to select only instances that are not a subclass of sth to gain the top level ones only.

poikilotherm commented:
I checked on the autocomplete/filtering support for controlled vocabulary fields in compound fields. Here's what I found:

1. For primitive fields using a CV we use a filter in case of "check multiple", but not for single values. This has been done as part of #6000 / PR #6339 (I knew there was an old issue for this... :smile: )
2. This change has not been introduced for "check multiple" in compound fields. No idea why. Tagging @mheppler here.
3. The remaining issue of single value fields has never been addressed, but @TaniaSchlatter [mentioned a view thoughts](https://github.com/IQSS/dataverse/issues/6000#issuecomment-509686007).

I guess adding the `filter` functionality to the "check multiple" fields in compound fields is an easy way forward (also I'd create a new issue for this, to keep this out of scope here). The fields where we would like to have benefit from this seem to be good fits for "check multiple" type. (So we might get around addressing number 3 from above)

As this seems like a good discussion for Dataverse software decoupled from this issue about CodeMeta, I'm going to create that little issue now.

mfenner commented:
@poikilotherm as Codemeta is close to version 3.0 (https://blog.datacite.org/codemeta-we-need-your-feedback/), `applicationCategory` and `scientific method` are good topics to discuss now. Would the Dataverse community want them to become part of Codemeta?

And what is the timing for this pull request with regards to Codemeta 2.0 vs. Codemeta 3.0 (which is still a few months away)?

poikilotherm commented:
> @poikilotherm as Codemeta is close to version 3.0 (https://blog.datacite.org/codemeta-we-need-your-feedback/), `applicationCategory` and `scientific method` are good topics to discuss now. Would the Dataverse community want them to become part of Codemeta?

@mfenner I think there is a high demand for these fields not only within the boundaries of the Dataverse community. I know that @sdruskat is also looking into this matter for his PhD thesis.

Are you aware of any existing, reusable controlled vocabularies, preferably as RDF/SKOS/JSON-LD/sth. with a PID, we could reuse for a field like `scientificMethod`? Dataverse soonish will have support to use those kind of sources within the UI (#7712)

> And what is the timing for this pull request with regards to Codemeta 2.0 vs. Codemeta 3.0 (which is still a few months away)?
I'm not so sure about this. Maybe it would be a good start to create a schema for 2.0 now and upgrade to 3.0 later on. It's a rather low hanging fruit. It might become necessary to introduce a migration method in Dataverse, but this seems like a good addition beyond the CodeMeta use case.

pdurbin commented:
@poikilotherm I couldn't get this tsv to load without making a few changes. I put them in a pull request for you to review and perhaps merge: https://github.com/poikilotherm/dataverse/pull/553

poikilotherm commented:
Thanks @pdurbin!

Just today I picked up working on this again (not yet pushed).

There's lots of stuff to be moved around, which will also incorporate your changes😉

poikilotherm commented on scripts/api/setup-datasetfields.sh:
```diff
@@ -7,3 +7,4 @@ curl http://localhost:8080/api/admin/datasetfield/load -X POST --data-binary @da
 curl http://localhost:8080/api/admin/datasetfield/load -X POST --data-binary @data/metadatablocks/astrophysics.tsv -H "Content-type: text/tab-separated-values"
 curl http://localhost:8080/api/admin/datasetfield/load -X POST --data-binary @data/metadatablocks/biomedical.tsv -H "Content-type: text/tab-separated-values"
 curl http://localhost:8080/api/admin/datasetfield/load -X POST --data-binary @data/metadatablocks/journals.tsv -H "Content-type: text/tab-separated-values"
+curl http://localhost:8080/api/admin/datasetfield/load -X POST --data-binary @data/metadatablocks/codemeta.tsv -H "Content-type: text/tab-separated-values"
```
**suggestion:** We might want to discuss to make this schema experimental, as we changed plans for #8812, too.

pdurbin replied:
Yes, good thought. Once we merge #8812 we could even follow up with a small PR to add Codemeta to the list of experimental blocks (in the Appendix of the User Guide, that is) and link here, if you want. That is, we could go ahead and advertise the work being done before putting the tsv itself into the code base. Same with other blocks being worked on: Darwin Core, 3D Object, etc. The idea is that people shouldn't start from scratch.

poikilotherm replied:
I merged the branch with latest `develop` and as #8812 is now merged, I moved the block to experimental, too.

poikilotherm commented:
Hi @doigl @atrisovic @4tikhonov @jggautier @pdurbin @adaybujeda @mmshad I just updated this PR with an improved block incorporating the latest news and decisions we had for this.

I also cleaned up the first comment to keep only the remaining questions (the old ones can still be found via the edit history). Please take a look. I guess we should discuss this on Community Slack and/or have a meeting of the WG again :smile:

atrisovic reviewed (commented):
LGTM!

jggautier commented:
Can we work on a plan for active user testing of the metadata block after it's merged? The appendix section of the guides says that feedback via any channel is welcome, but I'm worried that this isn't enough to move this from the new experimental label. (I have the same concerns about the computational workflow metadata block, but that's associated with a grant-funded project and I'm less sure of what the plans are for it.)

I think the [metadata block style guide](https://guides.dataverse.org/en/latest/style/text.html#metadata-text-guidelines) should be used to improve the labels and tooltip texts for this metadata block, too (and using it can help improve that style guide). Maybe this can be done in another step after it's been merged, like as part of another GitHub issue?

bast commented:
Thank you for your work on this! (sorry I know Dataverse only as a user ...)

- **Software license**: would this be selected/specified outside of these new forms? In the "normal" forms?
- What if a record/archive contains both data and code and the data has different license requirements (typically creative-commons) than software (typically not creative-commons). Would people be able to specify multiple licenses?

sdruskat commented:
>     * **Software license**: would this be selected/specified outside of these new forms? In the "normal" forms?
> 
>     * What if a record/archive contains both data and code and the data has different license requirements (typically creative-commons) than software (typically not creative-commons). Would people be able to specify multiple licenses?

@bast raises an important point here, something we face in the Citation File Format as well. The latter currently supports multiple licenses, albeit without scope. @schlauch suggested the use of [SPDX](https://spdx.org/licenses/) [expressions](https://spdx.github.io/spdx-spec/SPDX-license-expressions/). So for a mixed data + software deposit, this could be expressed as something like `MIT AND CC-BY-4.0`. License scope is hard to solve, but can be done in the deposit artifacts as well using, e.g., [REUSE](https://reuse.software/spec/).

poikilotherm commented:
> Can we work on a plan for active user testing of the metadata block after it's merged? The appendix section of the guides says that feedback via any channel is welcome, but I'm worried that this isn't enough to move this from the new experimental label.

@jggautier I hope to see this metadata block enabled on https://demo.dataverse.org! [We (project HERMES)](https://project.software-metadata.pub) are planning on creating [demo workflows](https://github.com/hermes-hmc/workflow/issues/59) with that instance and make good use of this metadata block!
 
> I think the [metadata block style guide](https://guides.dataverse.org/en/latest/style/text.html#metadata-text-guidelines) should be used to improve the labels and tooltip texts for this metadata block, too (and using it can help improve that style guide). Maybe this can be done in another step after it's been merged, like as part of another GitHub issue?

Absolutely. I wasn't aware of this guide, but let's use it to polish things. I love the idea of making incremental steps, but if you feel like creating a review with [suggestions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request), we could even incorporate changes upfront!

> * **Software license**: would this be selected/specified outside of these new forms? In the "normal" forms?

@bast this PR is about a metadata block. Licenses in Dataverse are a completely separate functionality and will not (and shall not) be affected by this PR.
 
> * What if a record/archive contains both data and code and the data has different license requirements (typically creative-commons) than software (typically not creative-commons). Would people be able to specify multiple licenses?

As @sdruskat layed out, this is a tricky thing on multiple different layers. I fear a solution to this is far beyond the possibilities and scope of this PR adding missing fields for CodeMeta support. You are very welcome to join the Dataverse Metadata Working Group meetings - we had been discussing #8512 last time which goes somewhat into the direction of your question.

poikilotherm commented:
Hi @jggautier!

> It makes sense that the metadata guidelines weren't considered. Work on this CodeMeta metadata block came way before these metadata guidelines were released.

Agreed! :wink: 

> I don't have time right now to apply the guidelines to the metadatablock and wouldn't want to hold up progress on it. So I agree it can be done incrementally, especially since the metadatablock is being labelled as experimental. These guidelines also need to be applied to other metadata fields that already ship with Dataverse, so I'll open a GitHub issue about that.

Sounds good to me!

> > @jggautier I hope to see this metadata block enabled on https://demo.dataverse.org/! [We (project HERMES)](https://project.software-metadata.pub/) are planning on creating [hermes-hmc/workflow#59](https://github.com/hermes-hmc/workflow/issues/59) with that instance and make good use of this metadata block!
> 
> So you hope to use Demo Dataverse with this metadatablock enabled on it in order to do testing with and collect feedback from depositors, curators, etc as part of project HERMES? Would you be able to share the results of this testing with the community? Then maybe the community could discuss what needs to be done for the metadata block to not be labelled as experimental.

Yes, we'd love to see that! It's always good to demonstrate workflows and it would be lovely if we can push to a "research software ready" Dataverse installation (and other can try it, too)!

@pdurbin  @mreekie I just added a release note docs to this PR and merged latest `develop`. As we have consensus of community (some folks reached out, no showstoppers), Metadata Working Group and IQSS, can we please go ahead and move this towards the SPRINT columns? It's small, very low risk, and was developed on HERMES funding. Thank you!

poikilotherm commented:
We did for the experimental workflows metadata.

pdurbin commented:
As discussed just now, the PR looks go overall.

We'd like to back out the schema.xml change.

(We'll make a PR to back out the schema.xml change for computational workflow as well, for consistency.)

Seems like @jggautier has given his blessing, especially since it's experimental.

Giving it a size of 10. This does not include deploying to demo.

poikilotherm commented:
@pdurbin @mreekie I just pushed the necessary changes to revert the addition to the schema. Also updated to latest `develop`. Dunno why the RTD CI fails, but seems unrelated.

poikilotherm commented:
> (We'll make a PR to back out the schema.xml change for computational workflow as well, for consistency.)

Chop chop here we go https://github.com/IQSS/dataverse/pull/9225

mreekie commented:
added to sprint Dec 15, 2022

pdurbin reviewed (commented):
Just a quick review. I haven't loaded up the block.

poikilotherm commented on src/main/java/propertyFiles/codeMeta20.properties:
```diff
@@ -0,0 +1,85 @@
+datasetfieldtype.softwareHelp.title=Software Help/Documentation
+datasetfieldtype.softwareHelp.description=Link to help texts or documentation
+datasetfieldtype.softwareHelp.watermark=e.g. https://user.github.io/project/docs
+datasetfieldtype.readme.title=Readme
```
I agree the filename should be sth with `README`. But do we want an all caps field name in the UI?

pdurbin replied:
Yes, I was trying to suggest all caps README in the UI. That's what you have in the description ("Link to the README of the project") and the watermark ("e.g. https://github.com/user/project/blob/main/README.md"), both of which appear in the UI, so it should probably be consistent, right?

It's weird, Codemeta itself has "link to software Readme file" as a description at https://codemeta.github.io/terms/ but before https://github.com/codemeta/codemeta/commit/0818c31 it was all caps README:

- before: "A URL for the software README file"
- after: "link to software Readme file"

pdurbin replied:
I haven't committed and pushed this, but here's how more consistent all uppercase README would look (name, description, and watermark):

![Screen Shot 2022-12-21 at 9 39 55 AM](https://user-images.githubusercontent.com/21006/208931405-02d5e30c-cc65-492c-9731-7d33c3e5756a.png)

poikilotherm commented on src/main/java/propertyFiles/codeMeta20.properties:
```diff
@@ -0,0 +1,85 @@
+datasetfieldtype.codeRepository.title=Code Repository
+datasetfieldtype.codeRepository.description=Link to the repository where the un-compiled, human readable code and related code is located (SVN, GitHub, CodePlex, institutional GitLab instance, Gitea, etc.).
+datasetfieldtype.codeRepository.watermark=e.g. https://github.com/user/project
+datasetfieldtype.applicationCategory.title=	Application Category
```
Done with f1a84b4

pdurbin reviewed (approved):
I played around with this locally and it's looking good!

I'm sending it to QA but I'll make a few observations:

- A lot of these fields would benefit from a picklist (programming languages, etc.) so I hope that we'll see a pull request to add some external controlled vocabularies.
- For the tooltips, there is some inconsistency of final periods being present or absent.
- It's weird that SVN is listed before Git, but that's because of CodeMeta and Schema.org.
- I'm slightly weirded out by the inconsistency between Readme (title) and README (tooltip and watermark).
- For some fields, it would be nice to have units (memory requirements, for example) but this is feedback to give upstream to CodeMeta and Schema.org, I imagine.
- I find "Target Product" to be a bit odd. Again, this is feedback to send upstream. I think the idea is that if, for example, you're creating a plugin for WordPress, you can put WordPress as the target product.
- There is a failing API test (FilesIT.test_008_ReplaceFileAlreadyDeleted) but I'm sure it has nothing to do with this metadata block, which isn't even loaded.
- It seems like Oliver would like more feedback earlier in the process. He posted about this at https://groups.google.com/g/dataverse-community/c/heNotzADbaQ/m/DJItrFjFBAAJ but in practice, developers like me don't take a serious look until the work (a PR in this case) make it into a sprint. So maybe we could improve our process here.

pdurbin commented:
@jggautier you probably missed the discussion but to sum up, only changes to non-experimental blocks should result in a change to schema.xml.

That is, schema.xml contains field for all the block that we ship. All these blocks are enabled by default and will "just work" because schema.xml has the fields already.

I hope this helps. This whole experimental blocks concept is quite new, of course!

jggautier commented:
Ah thanks. That's how I understood it. Experimental metadatablocks shouldn't be enabled in installations by default when those installations use the version of the software that includes that experimental metadatablock. Those installations will need to take extra steps to enable it.

It's just not clear to me how a metadatablock becomes not experimental.

pdurbin commented:
It hasn't happened yet! 😄 I hope we find out with CodeMeta!
